Duplicate playLevelView / refactor for Ozaria

AndrewJakubowicz opened the pull request:
The intent of this PR is to create a new base that we can build Ozaria on top of. Ideally we keep this folder entirely JavaScript, leveraging the linting work, testing work, vue work, and JsDocs.

I also moved the Cinematic Editor over to Ozaria attempting to adhere to the best practices to see how it all feels. It feels great!
The only exception was the cinematic.schema file which I left in the schema directory. This is due to the server dependencies on that folder specifically...

> Why not fork everything?

There is simply too much. Instead, I used our discussed and documented best practices to create a parallel `ozaria` folder with our new content!!!

## Ozaria Play Level View Refactor

I converted the play level view to javascript using [decaffeinate](https://github.com/decaffeinate/decaffeinate) and it worked very smoothly.

I was able to clean up a lot of the quirks of transpilation, and delete entire methods that have been inactive.

Two systems I identified and removed:
 - We used to identify machines that were strong enough and run arena simulations while they were playing. That has been turned off for a year.
 - Chat room system that hasn't been turned on.

Also deleted unused imports.

Some methods are still very verbose and ugly. I ran out of time cleaning them up, however they functionally work.

## Pain points I encountered

 - When using vueRouter, you're currently locked into the `base-flat` template.
 - We have a long chain of view inheritance that feels like to be simplified between the two new Vue components.
    - SingletonAppVueComponentView -> VueComponentView -> RootView -> CocoView
  - When using the vue router, how do we navigate between pages. The anchor tag wouldn't re-render the Vue component and state wouldn't be initialized. I believe the on-mount hooks stopped working. I'll need to look into this more.
  - Strangely the dynamic import chunk would not work as `ozariaPlay`... It started working when I used `OzariaPlay`.

## Other thoughts

How much more should we cut from the Play View?
The new vue components we write will have very predictable semantics, and will hopefully be tested and robust. How can we ensure we don't break underlying systems as we refactor and build on top of the existing codebase?

How do we tackle improvements on both CodeCombat and Ozaria?

Do we need a forked [OzariaVerifier](https://github.com/codecombat/codecombat/blob/88cda43cb9acbe981254b6072937b1ce2d748f14/app/views/editor/verifier/VerifierTest.coffee#L10)?

How do we keep on top of tests for quality during this fork?

Why have I left the decaffeinate comments on the top? I've left them there as they have not all been addressed to the fullest extent. This reminds us that this is a decaffeinated file. Eventually I could imagine removing this. Is it better to have a forked coffeescript file that's clean? Or a rougher javascript file that leverages the linting and editor improvements?

## To do list: 

   - [x] Put it behind an admin flag.
   - [x] Bring cinematics over to the ozaria folder following the Vue migration plan of folder structure.
   - [x] Delete obsolete code.
   - [x] Convert the forked playLevelView to JS.
   - [ ] See how many systems can be brought over and refactored. Potentially look into decoupling the Level Loading.

shubhi1092 commented on app/core/Router.coffee:
```diff
@@ -95,8 +95,6 @@ module.exports = class CocoRouter extends Backbone.Router
     'careers': => window.location.href = 'https://jobs.lever.co/codecombat'
     'Careers': => window.location.href = 'https://jobs.lever.co/codecombat'
 
-    'cinematic': go('CinematicView')
```
Will this route also be served from vuerouter?

AndrewJakubowicz replied:
Thanks for picking this up. I've completely removed this view while we figure out how to create the vue routes with different base templates.

jmif replied:
In order to do this we'll have to implement base-flat in Vue and then include it as needed (for example include it in school admin dashboard).  Looked at this briefly and we because we have the raw jade component we should be able to do this quickly.  We can discuss more in eng sync.

shubhi1092 commented on app/lib/dynamicRequire.js:
```diff
@@ -133,4 +131,7 @@ module.exports = {
   'views/core/SingletonAppVueComponentView': function () { return import(/* webpackChunkName: "vue" */ 'views/core/SingletonAppVueComponentView').then(m => m.default) },
+
+  // Ozaria file split
+  'views/ozaria/play/level/PlayLevelView': function () { return import(/* webpackChunkName: "OzariaPlay" */ 'ozaria/views/play/level/PlayLevelView') },
```
Its strange that `ozariaPlay` didnt work with a small `o`. OzariaUnitMap also has the same chunk name, I wonder how is that working.

AndrewJakubowicz replied:
I spent an hour trying to figure this out. I have a hypothesis that it's because of [chunk size](https://hackernoon.com/the-100-correct-way-to-split-your-chunks-with-webpack-f8a9df5b7758)? In reality I have no idea. It just wouldn't work until I made this change.

I'm unsure how to proceed.

differentmatt replied:
It's ok to note it as a mystery and proceed if it works otherwise.

shubhi1092 commented on ozaria/lib/cinematic/CameraSystem.js:
```diff
@@ -1,4 +1,4 @@
-import { getCamera } from '../../../schemas/selectors/cinematic'
+import { getCamera } from '../../schemas/selectors/cinematic'
```
Is camera system and cinematic lank boss meant to be shared across all of Ozaria. If not, should it be under `ozaria/components/cinematic/common` folder?

shubhi1092 replied:
Same for other files below in the ozaria/lib folder

AndrewJakubowicz replied:
My belief was that the components folder was specifically for `vue` files. The cinematic lank boss and other non ui cinematic systems are scoped within the `ozaria/lib/cinematic` folder. This is matching the pattern we use within app.
For example. The non-UI `world` system currently lives within `app/lib/world`. And aether is within `app/lib/aether`.

Do we want to put non UI systems in the components folder close to the vue views?
Or have a different directory structure rather than `ozaria/lib/<system>` for non UI systems?

shubhi1092 commented on ozaria/schemas/selectors/cinematic.js:
```diff
@@ -0,0 +1,417 @@
+import { CAMERA_DEFAULT } from '../../lib/cinematic/CameraSystem'
```
Is this file different from `/ozaria/schema/models/selectors/cinematic` ?

AndrewJakubowicz replied:
This was my error. I got confused about where we should keep files related to the schemas. Do we want to also split out new schemas into `ozaria`? Or should we keep them all in the legacy folder? Because the server depends on them I think I'm for keeping them in the legacy folder instead  of spreading that server dependency.

Really interested in other opinions.

jmif replied:
I'm +1 for keeping cinematic schema in schemas folder.  I think the schemas folder will be around for a while and I think cinematics can be considered a top level system that could (theoretically) span other games / pieces of the code base.

differentmatt replied:
Schemas in legacy folder seems fine.

shubhi1092 commented:
Reviewed the refactor in terms of folder structure, looks like ozaria folder mainly contains the following folders:
- `components` folder for vue components
- `lib` folder for utilities
- `templates`, `views` folder for backbone
- `schemas` folder
- `api` folder

Can you confirm if this is the correct understanding, and also describe the internal structure of `components` folder?

Also it looks like the `lib` folder should contain the helper files which are common across all of ozaria, would it make sense to move the cinematic helpers to `/components/cinematic` folder somewhere?

AndrewJakubowicz commented:
The only new folder that I added was the  `components` folder which I made from the https://codecombat.slab.com/drafts/vue-best-practices-pbb022cg#vue-layout.
Otherwise I've tried to keep the meaning of the other folders the same as in the legacy folder.  I've held back from writing down the ozaria  structure until after I get more feedback.

I made a different assumption regarding the location of the non-UI cinematic systems. I grouped them into the `lib` folder and didn't treat it as something that was global to Ozaria. You mentioned these could be better  grouped into a common folder within components but it didn't seem to fit the description of:

> common/ <-- Common components for teachers (or other system) section

that @jmif documented. I think I read that as literal `vue` components instead of common systems.

I'm now uncertain where to put the Ozaria specific systems. Because it's definitely not global so `lib` seems wrong. But it's not a shared vue component. But I feel I may have misunderstood the folders.

AndrewJakubowicz commented on app/core/vueRouter.js:
```diff
@@ -16,6 +18,11 @@ export default function getVueRouter () {
 
       routes: [
+        {
+          path: '/editor/cinematic/:slug?',
```
Does this nesting work if there is no component for the base `/editor` path yet?
I may punt this change, although it's an excellent suggestion.

AndrewJakubowicz commented on ozaria/components/cinematic/PageCinematicEditor/index.vue:
```diff
@@ -66,18 +20,22 @@ module.exports = Vue.extend({
   },
   mounted () {
-    if (!me.isAdmin()) {
+    if (!me.hasCinematicAccess()) {
```
Adding as a comment to add in a future iteration.

AndrewJakubowicz commented:
I've refactored this PR as of the Engineering Sync that we had today.

The major details are stored there. I will endeavor to document more of these decisions into slab as we move forward.

I also renamed many of the Cinematic system files to have better names.

Finally I added comments regarding using vue guards and the chunk name mystery.

How did I test these changes:

 - I ran my test cinematic that tests all the features.
 - I played `/ozaria/play/level/enemy-mine?` on an admin account and ensured it worked.
 - Checked that non admins cannot access `/ozaria/play/level/enemy-mine?`, as well as the `/editor/cinematic` route.

All other tests green.
